dpbench/infrastructure/benchmark.py

```diff
@@ -22,6 +26,138 @@
 from .numba_framework import NumbaFramework
 
 
+def _reset_output_args(bench, fmwrk, inputs, preset):
+    try:
+        output_args = bench.info["output_args"]
+    except KeyError:
+        logging.info(
+            "No output args to reset as benchmarks has no array output."
+        )
+        return
+    array_args = bench.info["array_args"]
+    for arg in inputs.keys():
+        if arg in output_args and arg in array_args:
+            original_data = bench.get_data(preset=preset)[arg]
+            inputs.update({arg: fmwrk.copy_to_func()(original_data)})
+
+
+def _exec(
+    bench,
+    fmwrk,
+    impl_postfix,
+    preset,
+    timeout,
+    repeat,
+    args,
+    results_dict,
+):
+    """Executes a benchmark for a given implementation.
+
+    A helper function to execute a benchmark. The function is called in a
+    separate sub-process by a BenchmarkRunner instance. The ``_exec`` function
+    first runs the benchmark implementation function once as a warmup and then
+    performs the specified number of repetitions. The output results are reset
+    before each repetition and the final output is serialized into a npz
+    (compressed NumPy data file) file.
+
+    All timing results and the path to the serialized results are written to
+    the results_dict input argument that is managed by the calling process.
+
+    Args:
+        bench : A Benchmark object representing the benchmark to be executed.
+        fmwrk : A Framework for which the benchmark is to be executed.
+        impl_postfix : The identifier for the benchmark implementation.
+        preset : A problem size entry defined in the bench_info JSON.
+        timeout : Number of seconds after which the execution is killed.
+        repeat : Number of repetitions of the benchmark execution.
+        args : Input arguments to benchmark implementation function.
+        results_dict : A dictionary where timing and other results are stored.
+    """
+    input_args = bench.info["input_args"]
+    array_args = bench.info["array_args"]
+    impl_fn = bench.get_impl(impl_postfix)
+    inputs = dict()
+    for arg in input_args:
+        if arg not in array_args:
+            inputs.update({arg: bench.get_data(preset=preset)[arg]})
+        else:
+            inputs.update({arg: args[arg]})
+
+    # Warmup
+    @tout.exit_after(timeout)
```

Comment:
Inside we are wrapping in timeout only warmup execution but join will check timeout for the entire execution. Also, if we set both timeouts to the same value the inner one most likely will never have chance to fire before process is killed.
So:
* Maybe wrap entire execution in `exit_after` instead of just warmup
* Give join timeout slightly more time (`timeout * 1.2`?)

Reply:
I have made the suggested changes. Please review once more.

Reply:
Is there any timeout handling remained except `join`? If no,  this `timeout * 1.2` probably is not needed. LGTM otherwise

Reply:
Nevermind, found `target=tout.exit_after(timeout)(_exec),`